### PR TITLE
fix: attention refactor and base doc

### LINKED ISSUE #229: Add conventions to triton kernel writing
Currently there are not conventions when writing triton kernel (naming for example)

### PR DESCRIPTION
**Pull request by gaetansnl**
Linear is partially refactored because input name is fixed by early_config_prune
I avoided important changes except refactor to linear:
- Removed
```
rematerialize rm and rn to save registers
rm = m_block_idx * BLOCK_M + tl.arange(0, BLOCK_M)
rn = n_block_idx * BLOCK_N + tl.arange(0, BLOCK_N)
```

- used variable initialized with tl.max_contiguous(tl.multiple_of everywhere

fixes #229

### CODE REVIEW

**Comment by gaetansnl** [comment]
full test passing

File: src/kernl/implementations/layer_norm.py [hunk 1]
```diff
@@ -59,16 +59,16 @@ def pytorch_naive_rmsnorm(a: torch.Tensor, weight: torch.Tensor, eps: float):
...
+    a_ptr,
+    weight_ptr,
+    bias_ptr,
+    mean_ptr,
+    rstd_ptr,
+    output_row_stride,
+    output_col_stride,
+    a_row_stride,
+    a_col_stride,
     N,
```
Comment (jonathlela) [inline src/kernl/implementations/layer_norm.py:72]: should we use `n_size` and `block_n_size` ?

File: src/kernl/implementations/linear_layer.py [hunk 2]
```diff
@@ -137,52 +137,53 @@ def kernel_fma(
...
     grid_n = (N + BLOCK_N - 1) // BLOCK_N
     # re-order program ID for better L2 performance
     width = GROUP_M * grid_n
-    group_id = pid // width
-    group_size = min(grid_m - group_id * GROUP_M, GROUP_M)
-    pid_m = group_id * GROUP_M + (pid % group_size)
-    pid_n = (pid % width) // (group_size)
+    group_idx = program_idx // width
+    group_size = min(grid_m - group_idx * GROUP_M, GROUP_M)
+    m_block_idx = group_idx * GROUP_M + (program_idx % group_size)
```
Comment (jonathlela) [inline src/kernl/implementations/linear_layer.py:148]: should it be `block_m_idx` and `block_n_idx` ?

File: src/kernl/implementations/attention_skinny.py [hunk 3]
```diff
@@ -88,48 +88,48 @@ def _fwd_part_1(
...
     IS_CAUSAL: tl.constexpr,
-    BLOCK_M: tl.constexpr,
-    BLOCK_DHEAD: tl.constexpr,
-    BLOCK_N: tl.constexpr,
+    BLOCK_M_SIZE: tl.constexpr,
+    BLOCK_DHEAD_SIZE: tl.constexpr,
+    BLOCK_N_SIZE: tl.constexpr,
 ):
     # Index of the block on M axis (M axis is the rows of matrix K)
     n_block_idx = tl.program_id(0)
```
Comment (jonathlela) [inline src/kernl/implementations/attention_skinny.py:114]: Should it be `block_n_idx` and `block_m_idx` ?

File: src/kernl/implementations/attention_skinny.py [hunk 4]
```diff
@@ -147,196 +147,200 @@ def _fwd_part_1(
     )
 
     # pointers to blocks in Q, K, V
-    ptrs_q = Q + offs_q
-    ptrs_k = K + offs_k
-    ptrs_v = V + offs_v
+    ptrs_q = q_ptr + offs_q
```
Comment (jonathlela) [inline src/kernl/implementations/attention_skinny.py:150]: `q_ptrs`, `k_ptrs` and `v_ptrs` ?

**Review by jonathlela — APPROVED** [review]
It's a bit hard to review but the new naming convention make the code easier to read.
There is still some omissions in the whole renaming. I've pointed out 3-4 of them in the feedback, not sure if there is others remaining.

File: src/kernl/implementations/attention_skinny.py (hunk 4 above)
Comment (pommedeterresautee) [reply src/kernl/implementations/attention_skinny.py]: image

File: src/kernl/implementations/README.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
+# Naming conventions
```
Comment (pommedeterresautee) [inline src/kernl/implementations/README.md:1]: README in a more visible location like contribution folder? With a good title / name so we know its purpose.

File: src/kernl/implementations/attention.py [hunk 6]
```diff
@@ -186,21 +186,21 @@ def _fwd_kernel(
                    ┌────────────┐
                │   │            │
     M Dimension│   ├────────────┤     ┌───┐
-      size_m   │   │            │     │   │ BLOCK_M
+      m_size   │   │            │     │   │ BLOCK_M
```
Comment (pommedeterresautee) [inline src/kernl/implementations/attention.py:189]: BLOCK_M_SIZE ? same fo N

File: src/kernl/implementations/attention_skinny.py (hunk 4 above)
Comment (gaetansnl) [reply src/kernl/implementations/attention_skinny.py]: done

File: src/kernl/implementations/attention_skinny.py (hunk 3 above)
Comment (gaetansnl) [reply src/kernl/implementations/attention_skinny.py]: done

File: src/kernl/implementations/linear_layer.py (hunk 2 above)
Comment (gaetansnl) [reply src/kernl/implementations/linear_layer.py]: done

File: src/kernl/implementations/layer_norm.py (hunk 1 above)
Comment (gaetansnl) [reply src/kernl/implementations/layer_norm.py]: done

File: src/kernl/implementations/README.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+- `_range_offs` output of `tl.arange(0, N)` to be used as offset for pointers
+-  `_ptr_mask` bool matrix or vector to be used at mask for load and store operations
+- `_LOAD_MASK_NEEDED` boolean, usually used as constexpr to determine if we need to use a mask when loading or storing tensor
+## Variable prefixes
+- `is_`, `has_` or `should_` for booleans
+
+## Dimension naming
+- Dimension is singular
+- If dimension follows variable name from a formula. You can use this name. Example MNK for matmul
+- Use `col` or `row` singular if you don't have a new for the last two dimensions
```
[the remark this answers is not in the document]
Comment (gaetansnl) [reply src/kernl/implementations/README.md]: done

File: src/kernl/implementations/README.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+## Variable suffixes
+We prefer suffixes to prefixes because
+- avoids looking at the middle of the variable to find the interesting part. 
+- It allows quick comparaison, for example `q = q_ptr + q_offs`, everything starts we `q` and we reduce mistakes.
+- It follows the pronunciation
+
+Suffixes list:
+- `_idx` integer representing an index
+- `_size` integer representing a size
+- `_off` integer reoresenting an offset
```
[the remark this answers is not in the document]
Comment (gaetansnl) [reply src/kernl/implementations/README.md]: done

**Review by pommedeterresautee — COMMENTED** [review]
Tried to review but it's really hard. 
Did it passed tests? (full pytest)
Added few remarks on top of those of @jonathlela

**Comment by gaetansnl** [comment]
@pommedeterresautee no got an error at the end, I will try to reproduce

**Review by white-gorilla — CHANGES_REQUESTED** [review]
The Markown is optimized but the lack of line breaks makes it less readable in `.md` for maintainers and not readable once converted to HTML on the site.

- Don't hesitate to put line breaks after the titles and especially before the lists (otherwise they will be displayed inline).
- Tip: stay consistent on punctuation, especially the `.` at the end of the line and the `:` (either you put them everywhere or you don't).

Otherwise it's fine.

Capture d’écran 2023-01-17 à 14 01 17

Capture d’écran 2023-01-17 à 13 58 43

Capture d’écran 2023-01-17 à 14 00 00

**Review by pommedeterresautee — APPROVED** [review]
quite hard to review line by line but from what I checked, it's ok.
plus test pass

**Comment by white-gorilla** [comment]
@gaetansnl Still missing a line break.

Capture d’écran 2023-01-17 à 17 01 37

**Review by white-gorilla — APPROVED** [review]

**Pull request merged**